### PR TITLE
cmake: add LDAP Find module

### PR DESCRIPTION
**Pull request by vszakats**
Move LDAP detection to its own Find module.

It supports `pkg-config` and the standard detection method used for
other dependencies, with version detection.

In curl CI it fixes LDAP detection in the OpenBSD job.

---

w/o whitespace: https://github.com/curl/curl/pull/15273/files?w=1

- [x] rebase on #15573.

### CODE REVIEW

**Comment by vszakats** [comment]
The native cmake detection gets weird and detects LDAP in the Apple SDK. That provides an `ldap.h` and `lber.h`, which are stubs including themselves (using `<ldap.h>`). This causes an infinite inclusion loop when compiling.

For now I opted out `APPLE` from using the native cmake detection, but this requires to keep the old manual LDAP detection logic, while it would be reasonable goal to drop that.

**Comment by vszakats** [comment]
After this patch, in CI, OpenLDAP is detected in the same jobs as before.
On OpenBSD it wasn't detected because the installation failed. Tackling this in #15491.

File: CMake/FindLDAP.cmake [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+   NOT DEFINED LDAP_LIBRARY AND
+   NOT DEFINED LDAP_LBER_LIBRARY)
+  find_package(PkgConfig QUIET)
+  pkg_check_modules(LDAP "ldap")
+  pkg_check_modules(LDAP_LBER "lber")
+endif()
+
+if(LDAP_FOUND AND LDAP_LBER_FOUND)
+  list(APPEND LDAP_LIBRARIES ${LDAP_LBER_LIBRARIES})
+  list(REMOVE_DUPLICATES LDAP_LIBRARIES)
```
Comment (dg0yt) [inline CMake/FindLDAP.cmake:52]: "Traditional" linkers rely on libary order. `REMOVE_DUPLICATES` would remove the late duplicates (e.g. the last `-lm`), and this breaks linking with "traditional" linkers which rely on lib order.

```suggestion
  list(REMOVE_ITEM LDAP_LIBRARIES ${LDAP_LBER_LIBRARIES})
  list(APPEND LDAP_LIBRARIES ${LDAP_LBER_LIBRARIES})
```

Comment (vszakats) [reply CMake/FindLDAP.cmake:52]: Note to self that this pattern is also used for mbedtls and brotli (pending PR).

And also for Libs.private in libcurl.pc.

I guess these can be dropped. Or come up with a function dropping the early dupes.

(There is also the "overlinking" issue when linking shared libs with these deps. Could't find a solution for that.)

Comment (vszakats) [reply CMake/FindLDAP.cmake:52]: Do you think this'd work?:
```
list(REVERSE LDAP_LIBRARIES)
list(REMOVE_DUPLICATES LDAP_LIBRARIES)
list(REVERSE LDAP_LIBRARIES)
```

Comment (dg0yt) [reply CMake/FindLDAP.cmake]: I used to use this pattern, too.

In general, I just leave the duplicates untouched today (and patch out `REMOVE_DUPLICATES`). Different linkers may need different strategies (accoding to CMake). With Apple frameworks, you may accidentally remove `-framework`.

**Pull request closed without merging**